graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java

```diff
@@ -20,11 +41,75 @@ protected MutationFetcher(
   }
 
   @Override
-  protected Map<String, Object> get(DataFetchingEnvironment environment, DataStore dataStore)
-      throws Exception {
-    String statement = buildStatement(environment, dataStore);
-    dataStore.query(statement).get();
-    return ImmutableMap.of("value", environment.getArgument("value"));
+  protected CompletableFuture<Map<String, Object>> get(
+      DataFetchingEnvironment environment, DataStore dataStore) {
+    String statement = null;
+    Exception buildException = null;
+
+    // Avoid mixing sync and async exceptions
+    try {
+      // buildStatement() could throw an unchecked exception.
+      // As the statement might be part of a batch, we need to make sure the
+      // batched operation completes.
+      statement = buildStatement(environment, dataStore);
+    } catch (Exception e) {
+      buildException = e;
+    }
+
+    OperationDefinition operation = environment.getOperationDefinition();
+
+    if (operation.getDirectives().stream().anyMatch(d -> d.getName().equals(ATOMIC_DIRECTIVE))
+        && operation.getSelectionSet().getSelections().size() > 1) {
+      // There are more than one mutation in @atomic operation
+      return executeAsBatch(environment, dataStore, statement, buildException);
+    }
+
+    if (buildException != null) {
+      CompletableFuture<Map<String, Object>> f = new CompletableFuture<>();
+      f.completeExceptionally(buildException);
+      return f;
+    }
+
+    // Execute as a single statement
+    return dataStore
+        .query(statement)
+        .thenApply(rs -> ImmutableMap.of("value", environment.getArgument("value")));
+  }
+
+  private CompletableFuture<Map<String, Object>> executeAsBatch(
+      DataFetchingEnvironment environment,
+      DataStore dataStore,
+      String statement,
+      Exception buildException) {
+    int selections = environment.getOperationDefinition().getSelectionSet().getSelections().size();
+    HTTPAwareContextImpl context = environment.getContext();
+    HTTPAwareContextImpl.BatchContext batchContext = context.getBatchContext();
+
+    if (environment.getArgument("options") != null) {
+      // Users should specify query options once in the batch
+      int counter = batchContext.setDataStore(dataStore);
+
+      if (counter > 1) {
+        // DataStore can be set at most once.
+        // The instance used is the one that came from the user options (if any).
+        buildException =
+            new GraphQLException(
+                "options can only de defined once in an @atomic mutation selection");
+      }
+    }
+
+    if (buildException != null) {
```

Comment:
with the current logic, if there is one `buildException` we are failing completing exceptionally with only one exception, right?
Shouldn't we include all possible exceptions, add those as suppressed one and complete exceptionally with all details?

Reply:
Good question.
In GraphQL, each mutation "selection", should have an individual response. As this is all or nothing, we fail all the selections at the same time with the same error.

Otherwise, it might lead the user to think that one succeeded and the others don't.